TimedArray with wrong argument number in Synapses connect condition fails ungraciously

I was just testing if `TimedArray` can also be used in a Synapse `connect` condition (which runs on the host for brian2CUDA). And it seems that is possible. But while a `TimedArray` in a `NeuronGroup` fails with a useful error message when the function is called with two few or two many arguments, this does not happen in the Synapse connect condition. Just took some example from the tests and adapted it below by a Synapse. This runs without error.
 
```python
from brian2 import *
set_device('cpp_standalone')

ta2d = TimedArray(np.arange(12).reshape(4, 3), dt=0.1*ms)
G = NeuronGroup(3, '''x : 1
                      value = ta2d(t, i) + 1: 1''',
                      dt=0.1*ms, threshold='True')
S = Synapses(G, G, on_pre='x += 1')
S.connect('ta2d(t, 2) > 3')
mon = StateMonitor(G, 'value', record=True, dt=0.1*ms)
run(0.5*ms)
```

If I now change the `NeuronGroup` equation to `value = ta2d(t) + 1: 1` (only one argument for 2d function), I get this nice error:
```python
SyntaxError: Function ta2d was called with 1 parameters, needs 2.
```

But if I change the Synapse connect condition to `ta2d(t) > 3`, I don't. Instead it fails ungraciously during the compilation with:
```cpp
|| code_objects/synapses_synapses_create_generator_codeobject.cpp: In function ‘void _run_synapses_synapses_create_generator_codeobject()’:
|| code_objects/synapses_synapses_create_generator_codeobject.cpp:136:50: error: too few arguments to function ‘double {anonymous}::_timedarray(double, int)’
||   136 |     const char _lio_create_cond_1 = _timedarray(t) > 3;
||       |                                                  ^
|| code_objects/synapses_synapses_create_generator_codeobject.cpp:20:23: note: declared here
||    20 |  static inline double _timedarray(const double t, const int i)
||       |                       ^~~~~~~~~~~
|| make: *** [makefile:31: code_objects/synapses_synapses_create_generator_codeobject.o] Error 1
|| make: *** Waiting for unfinished jobs....
```

Not sure if it makes sense to have `TimedArray` in a codeobject that does not run on a clock? But if it's possible, I guess the error should be shown there as well :) Or not allow it at all?

--- a/brian2/parsing/expressions.py
+++ b/brian2/parsing/expressions.py
@@ -235,7 +235,7 @@ def parse_expression_dimensions(expr, variables):
             raise SyntaxError('%s was used like a variable/constant, but it is '
                               'a function.' % name)
         if name in variables:
-            return variables[name].dim
+            return get_dimensions(variables[name])
         elif name in ['True', 'False']:
             return DIMENSIONLESS
         else:

--- a/brian2/parsing/expressions.py
+++ b/brian2/parsing/expressions.py
@@ -190,7 +190,7 @@ def _get_value_from_expression(expr, variables):
         raise SyntaxError('Unsupported operation ' + str(expr.__class__))
 
     
-def parse_expression_dimensions(expr, variables):
+def parse_expression_dimensions(expr, variables, orig_expr=None):
     '''
     Returns the unit value of an expression, and checks its validity
     
@@ -218,6 +218,7 @@ def parse_expression_dimensions(expr, variables):
 
     # If we are working on a string, convert to the top level node    
     if isinstance(expr, str):
+        orig_expr = expr
         mod = ast.parse(expr, mode='eval')
         expr = mod.body
     if expr.__class__ is getattr(ast, 'NameConstant', None):
@@ -233,7 +234,12 @@ def parse_expression_dimensions(expr, variables):
         # (most of the time this happens for a TimedArray)
         if name in variables and isinstance(variables[name], Function):
             raise SyntaxError('%s was used like a variable/constant, but it is '
-                              'a function.' % name)
+                              'a function.' % name,
+                              ("<string>",
+                               expr.lineno,
+                               expr.col_offset + 1,
+                               orig_expr)
+                              )
         if name in variables:
             return get_dimensions(variables[name])
         elif name in ['True', 'False']:
@@ -246,7 +252,7 @@ def parse_expression_dimensions(expr, variables):
     elif expr.__class__ is ast.BoolOp:
         # check that the units are valid in each subexpression
         for node in expr.values:
-            parse_expression_dimensions(node, variables)
+            parse_expression_dimensions(node, variables, orig_expr=orig_expr)
         # but the result is a bool, so we just return 1 as the unit
         return DIMENSIONLESS
     elif expr.__class__ is ast.Compare:
@@ -254,7 +260,7 @@ def parse_expression_dimensions(expr, variables):
         subexprs = [expr.left]+expr.comparators
         subunits = []
         for node in subexprs:
-            subunits.append(parse_expression_dimensions(node, variables))
+            subunits.append(parse_expression_dimensions(node, variables, orig_expr=orig_expr))
         for left_dim, right_dim in zip(subunits[:-1], subunits[1:]):
             if not have_same_dimensions(left_dim, right_dim):
                 msg = ('Comparison of expressions with different units. Expression '
@@ -274,7 +280,12 @@ def parse_expression_dimensions(expr, variables):
 
         func = variables.get(expr.func.id, None)
         if func is None:
-            raise SyntaxError('Unknown function %s' % expr.func.id)
+            raise SyntaxError('Unknown function %s' % expr.func.id,
+                              ("<string>",
+                               expr.lineno,
+                               expr.col_offset + 1,
+                               orig_expr)
+                              )
         if not hasattr(func, '_arg_units') or not hasattr(func, '_return_unit'):
             raise ValueError(('Function %s does not specify how it '
                               'deals with units.') % expr.func.id)
@@ -283,7 +294,11 @@ def parse_expression_dimensions(expr, variables):
             raise SyntaxError('Function %s was called with %d parameters, '
                               'needs %d.' % (expr.func.id,
                                              len(expr.args),
-                                             len(func._arg_units)))
+                                             len(func._arg_units)),
+                              ("<string>",
+                               expr.lineno,
+                               expr.col_offset + len(expr.func.id) + 1,
+                               orig_expr))
 
         for idx, (arg, expected_unit) in enumerate(zip(expr.args,
                                                        func._arg_units)):
@@ -297,7 +312,7 @@ def parse_expression_dimensions(expr, variables):
                                      '"%s".') % (idx + 1, expr.func.id,
                                                  NodeRenderer().render_node(arg)))
             else:
-                arg_unit = parse_expression_dimensions(arg, variables)
+                arg_unit = parse_expression_dimensions(arg, variables, orig_expr=orig_expr)
                 if not have_same_dimensions(arg_unit, expected_unit):
                     msg = ('Argument number {} for function {} does not have the '
                            'correct units. Expression "{}" has units ({}), but '
@@ -314,14 +329,14 @@ def parse_expression_dimensions(expr, variables):
             return getattr(func._return_unit, 'dim', DIMENSIONLESS)
         else:
             # Function returns a unit that depends on the arguments
-            arg_units = [parse_expression_dimensions(arg, variables)
+            arg_units = [parse_expression_dimensions(arg, variables, orig_expr=orig_expr)
                          for arg in expr.args]
             return func._return_unit(*arg_units).dim
 
     elif expr.__class__ is ast.BinOp:
         op = expr.op.__class__.__name__
-        left_dim = parse_expression_dimensions(expr.left, variables)
-        right_dim = parse_expression_dimensions(expr.right, variables)
+        left_dim = parse_expression_dimensions(expr.left, variables, orig_expr=orig_expr)
+        right_dim = parse_expression_dimensions(expr.right, variables, orig_expr=orig_expr)
         if op in ['Add', 'Sub', 'Mod']:
             # dimensions should be the same
             if left_dim is not right_dim:
@@ -346,8 +361,17 @@ def parse_expression_dimensions(expr, variables):
             u = left_dim/right_dim
         elif op == 'FloorDiv':
             if not (left_dim is DIMENSIONLESS and right_dim is DIMENSIONLESS):
+                if left_dim is DIMENSIONLESS:
+                    col_offset = expr.right.col_offset + 1
+                else:
+                    col_offset = expr.left.col_offset + 1
                 raise SyntaxError('Floor division can only be used on '
-                                  'dimensionless values.')
+                                  'dimensionless values.',
+                                  ("<string>",
+                                   expr.lineno,
+                                   col_offset,
+                                   orig_expr)
+                                  )
             u = DIMENSIONLESS
         elif op == 'Pow':
             if left_dim is DIMENSIONLESS and right_dim is DIMENSIONLESS:
@@ -355,15 +379,25 @@ def parse_expression_dimensions(expr, variables):
             n = _get_value_from_expression(expr.right, variables)
             u = left_dim**n
         else:
-            raise SyntaxError("Unsupported operation "+op)
+            raise SyntaxError("Unsupported operation "+op,
+                              ("<string>",
+                               expr.lineno,
+                               expr.left.end_col_offset + 1,
+                               orig_expr)
+                              )
         return u
     elif expr.__class__ is ast.UnaryOp:
         op = expr.op.__class__.__name__
         # check validity of operand and get its unit
-        u = parse_expression_dimensions(expr.operand, variables)
+        u = parse_expression_dimensions(expr.operand, variables, orig_expr=orig_expr)
         if op == 'Not':
             return DIMENSIONLESS
         else:
             return u
     else:
-        raise SyntaxError('Unsupported operation ' + str(expr.__class__))
+        raise SyntaxError('Unsupported operation ' + str(expr.__class__.__name__),
+                          ("<string>",
+                           expr.lineno,
+                           expr.col_offset + 1,
+                           orig_expr)
+                          )

--- a/brian2/parsing/expressions.py
+++ b/brian2/parsing/expressions.py
@@ -382,7 +382,8 @@ def parse_expression_dimensions(expr, variables, orig_expr=None):
             raise SyntaxError("Unsupported operation "+op,
                               ("<string>",
                                expr.lineno,
-                               expr.left.end_col_offset + 1,
+                               getattr(expr.left, 'end_col_offset',
+                                       len(NodeRenderer().render_node(expr.left))) + 1,
                                orig_expr)
                               )
         return u

--- a/brian2/synapses/synapses.py
+++ b/brian2/synapses/synapses.py
@@ -25,6 +25,7 @@ from brian2.groups.group import Group, CodeRunner, get_dtype
 from brian2.groups.neurongroup import (extract_constant_subexpressions,
                                        SubexpressionUpdater,
                                        check_identifier_pre_post)
+from brian2.parsing.expressions import is_boolean_expression, parse_expression_dimensions
 from brian2.stateupdaters.base import (StateUpdateMethod,
                                        UnsupportedEquationsException)
 from brian2.stateupdaters.exact import linear, independent
@@ -1311,10 +1312,14 @@ class Synapses(Group):
             raise ValueError("Generator expression given for condition, write "
                              "connect(j='{condition}'...) instead of "
                              "connect('{condition}'...).".format(condition=condition))
-        # TODO: check if string expressions have the right types and return
-        # useful error messages
 
         self._connect_called = True
+
+        # Get namespace information and merge it with the variables
+        if namespace is None:
+            namespace = get_local_namespace(level=level + 2)
+        variables = dict(namespace)
+        variables.update(self.variables)
 
         # which connection case are we in?
         if condition is None and i is None and j is None:
@@ -1329,6 +1334,18 @@ class Synapses(Group):
                     return
                 if condition is True:
                     condition = 'True'
+                # Check that the condition is a boolean expresion
+                if not is_boolean_expression(condition, variables):
+                    raise TypeError(f'Condition \'{condition}\' is not a '
+                                    f'boolean condition')
+
+                # Check the units (mostly to check for unit consistency within the condition)
+                dims = parse_expression_dimensions(condition, variables)
+                if dims is not DIMENSIONLESS:
+                    # We should not get here normally
+                    raise TypeError(f'Condition \'{condition}\' is not a '
+                                    f'boolean condition')
+
                 condition = word_substitute(condition, {'j': '_k'})
                 if not isinstance(p, str) and p == 1:
                     j = ('_k for _k in range(N_post) '
@@ -1336,8 +1353,6 @@ class Synapses(Group):
                 else:
                     j = None
                     if isinstance(p, str):
-                        if namespace is None:
-                            namespace = get_local_namespace(level=level + 1)
                         p_dep = self._expression_index_dependence(p, namespace=namespace)
                         if '_postsynaptic_idx' in p_dep or '_iterator_idx' in p_dep:
                             j = ('_k for _k in range(N_post) '

--- a/brian2/synapses/synapses.py
+++ b/brian2/synapses/synapses.py
@@ -29,7 +29,7 @@ from brian2.parsing.expressions import is_boolean_expression, parse_expression_d
 from brian2.stateupdaters.base import (StateUpdateMethod,
                                        UnsupportedEquationsException)
 from brian2.stateupdaters.exact import linear, independent
-from brian2.units.fundamentalunits import (Quantity, DIMENSIONLESS,
+from brian2.units.fundamentalunits import (Quantity, DIMENSIONLESS, DimensionMismatchError,
                                            fail_for_dimension_mismatch)
 from brian2.units.allunits import second
 from brian2.utils.logger import get_logger
@@ -1353,6 +1353,9 @@ class Synapses(Group):
                 else:
                     j = None
                     if isinstance(p, str):
+                        dim = parse_expression_dimensions(p, variables)
+                        if dim is not DIMENSIONLESS:
+                            raise DimensionMismatchError('Expression for p should be dimensionless.')
                         p_dep = self._expression_index_dependence(p, namespace=namespace)
                         if '_postsynaptic_idx' in p_dep or '_iterator_idx' in p_dep:
                             j = ('_k for _k in range(N_post) '

--- a/brian2/synapses/synapses.py
+++ b/brian2/synapses/synapses.py
@@ -1315,11 +1315,9 @@ class Synapses(Group):
 
         self._connect_called = True
 
-        # Get namespace information and merge it with the variables
+        # Get namespace information
         if namespace is None:
             namespace = get_local_namespace(level=level + 2)
-        variables = dict(namespace)
-        variables.update(self.variables)
 
         # which connection case are we in?
         if condition is None and i is None and j is None:
@@ -1335,6 +1333,8 @@ class Synapses(Group):
                 if condition is True:
                     condition = 'True'
                 # Check that the condition is a boolean expresion
+                identifiers = get_identifiers(condition)
+                variables = self.resolve_all(identifiers, namespace)
                 if not is_boolean_expression(condition, variables):
                     raise TypeError(f'Condition \'{condition}\' is not a '
                                     f'boolean condition')
@@ -1353,6 +1353,8 @@ class Synapses(Group):
                 else:
                     j = None
                     if isinstance(p, str):
+                        identifiers = get_identifiers(p)
+                        variables = self.resolve_all(identifiers, namespace)
                         dim = parse_expression_dimensions(p, variables)
                         if dim is not DIMENSIONLESS:
                             raise DimensionMismatchError('Expression for p should be dimensionless.')

--- a/brian2/tests/test_parsing.py
+++ b/brian2/tests/test_parsing.py
@@ -245,67 +245,72 @@ def test_is_boolean_expression():
 
 
 @pytest.mark.codegen_independent
-def test_parse_expression_unit():
+@pytest.mark.parametrize('expect,expr', [
+    (volt*amp, 'a+b*c'),
+    (DimensionMismatchError, 'a+b'),
+    (DimensionMismatchError, 'a<b'),
+    (1, 'a<b*c'),
+    (1, 'a or b'),
+    (1, 'not (a >= b*c)'),
+    (DimensionMismatchError, 'a or b<c'),
+    (1, 'a/(b*c)<1'),
+    (1, 'a/(a-a)'),
+    (1, 'a<mV*mA'),
+    (volt**2, 'b**2'),
+    (volt*amp, 'a%(b*c)'),
+    (volt, '-b'),
+    (1, '(a/a)**(a/a)'),
+    # Expressions involving functions
+    (volt, 'rand()*b'),
+    (volt**0.5, 'sqrt(b)'),
+    (volt, 'ceil(b)'),
+    (volt, 'sqrt(randn()*b**2)'),
+    (1, 'sin(b/b)'),
+    (DimensionMismatchError, 'sin(b)'),
+    (DimensionMismatchError, 'sqrt(b) + b'),
+    (SyntaxError, 'sqrt(b, b)'),
+    (SyntaxError, 'sqrt()'),
+    (SyntaxError, 'int(1, 2)'),
+])
+def test_parse_expression_unit(expect, expr):
     Var = namedtuple('Var', ['dim', 'dtype'])
     variables = {'a': Var(dim=(volt*amp).dim, dtype=np.float64),
                  'b': Var(dim=volt.dim, dtype=np.float64),
                  'c': Var(dim=amp.dim, dtype=np.float64)}
     group = SimpleGroup(namespace={}, variables=variables)
-    EE = [
-        (volt*amp, 'a+b*c'),
-        (DimensionMismatchError, 'a+b'),
-        (DimensionMismatchError, 'a<b'),
-        (1, 'a<b*c'),
-        (1, 'a or b'),
-        (1, 'not (a >= b*c)'),
-        (DimensionMismatchError, 'a or b<c'),
-        (1, 'a/(b*c)<1'),
-        (1, 'a/(a-a)'),
-        (1, 'a<mV*mA'),
-        (volt**2, 'b**2'),
-        (volt*amp, 'a%(b*c)'),
-        (volt, '-b'),
-        (1, '(a/a)**(a/a)'),
-        # Expressions involving functions
-        (volt, 'rand()*b'),
-        (volt**0.5, 'sqrt(b)'),
-        (volt, 'ceil(b)'),
-        (volt, 'sqrt(randn()*b**2)'),
-        (1, 'sin(b/b)'),
-        (DimensionMismatchError, 'sin(b)'),
-        (DimensionMismatchError, 'sqrt(b) + b'),
-        (SyntaxError, 'sqrt(b, b)'),
-        (SyntaxError, 'sqrt()'),
-        (SyntaxError, 'int(1, 2)'),
-        ]
-    for expect, expr in EE:
-        all_variables = {}
-        for name in get_identifiers(expr):
-            if name in variables:
-                all_variables[name] = variables[name]
-            else:
-                all_variables[name] = group._resolve(name, {})
-
-        if isinstance(expect, type) and issubclass(expect, Exception):
-            with pytest.raises(expect):
-                parse_expression_dimensions(expr, all_variables)
+    all_variables = {}
+    for name in get_identifiers(expr):
+        if name in variables:
+            all_variables[name] = variables[name]
         else:
-            u = parse_expression_dimensions(expr, all_variables)
-            assert have_same_dimensions(u, expect)
+            all_variables[name] = group._resolve(name, {})
 
-    wrong_expressions = ['a**b',
-                         'a << b',
-                         'int(True' # typo
-                        ]
-    for expr in wrong_expressions:
-        all_variables = {}
-        for name in get_identifiers(expr):
-            if name in variables:
-                all_variables[name] = variables[name]
-            else:
-                all_variables[name] = group._resolve(name, {})
-        with pytest.raises(SyntaxError):
+    if isinstance(expect, type) and issubclass(expect, Exception):
+        with pytest.raises(expect):
             parse_expression_dimensions(expr, all_variables)
+    else:
+        u = parse_expression_dimensions(expr, all_variables)
+        assert have_same_dimensions(u, expect)
+
+@pytest.mark.codegen_independent
+@pytest.mark.parametrize('expr', ['a**b',
+                           'a << b',
+                           'int(True' # typo
+                           ])
+def test_parse_expression_unit_wrong_expressions(expr):
+    Var = namedtuple('Var', ['dim', 'dtype'])
+    variables = {'a': Var(dim=(volt*amp).dim, dtype=np.float64),
+                 'b': Var(dim=volt.dim, dtype=np.float64),
+                 'c': Var(dim=amp.dim, dtype=np.float64)}
+    all_variables = {}
+    group = SimpleGroup(namespace={}, variables=variables)
+    for name in get_identifiers(expr):
+        if name in variables:
+            all_variables[name] = variables[name]
+        else:
+            all_variables[name] = group._resolve(name, {})
+    with pytest.raises(SyntaxError):
+        parse_expression_dimensions(expr, all_variables)
 
 
 @pytest.mark.codegen_independent

--- a/brian2/tests/test_synapses.py
+++ b/brian2/tests/test_synapses.py
@@ -94,6 +94,14 @@ def test_connect_errors():
     with pytest.raises(SyntaxError):
         S.connect('sin(3, 4) > 1')
 
+    # Unit error in p argument
+    with pytest.raises(TypeError):
+        S.connect('1*mV')
+
+    # Syntax error in p argument
+    with pytest.raises(SyntaxError):
+        S.connect(p='sin(3, 4)')
+
 
 @pytest.mark.codegen_independent
 def test_name_clashes():

--- a/brian2/tests/test_synapses.py
+++ b/brian2/tests/test_synapses.py
@@ -76,6 +76,25 @@ def test_creation_errors():
     with pytest.raises(TypeError):
         Synapses(G, G, 'w:1', post='v+=w', on_post='v+=w', connect=True)
 
+
+@pytest.mark.codegen_independent
+def test_connect_errors():
+    G = NeuronGroup(42, '')
+    S = Synapses(G, G)
+
+    # Not a boolean condition
+    with pytest.raises(TypeError):
+        S.connect('i*2')
+
+    # Unit error
+    with pytest.raises(DimensionMismatchError):
+        S.connect('i > 3*mV')
+
+    # Syntax error
+    with pytest.raises(SyntaxError):
+        S.connect('sin(3, 4) > 1')
+
+
 @pytest.mark.codegen_independent
 def test_name_clashes():
     # Using identical names for synaptic and pre- or post-synaptic variables

